### PR TITLE
ThreatQ Integration

### PR DESCRIPTION
**Pull request by dantavori**
## Status
In Progress

## Related Issues
https://github.com/demisto/etc/issues/6677

## Description
ThreatQ provides indicator scoring weighting for indicators and their contextual information, such as sources, attributes, and indicator types, as they are added to ThreatQ.
For detailed information on ThreatQ scoring please refer to https://helpcenter.threatq.com/

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
**Added the following commands:**
tq-search-by-name
ip
url
file
email
domain
tq-create-ioc
tq-add-attr
tq-modify-attr
tq-link-objects
tq-create-adversary
tq-create-event
tq-get-related-events
tq-get-related-adversaries
tq-upload-file
tq-search-by-id
tq-unlink-objects
tq-delete-object
tq-add-source
tq-delete-source
tq-delete-attr

**TODO:**
tq-upload-file

### CODE REVIEW

**Comment by Itay4** [comment]
command names prefix  should be `threatq-`

File: Integrations/ThreatQ/ThreatQ.py [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+    try:
+        file_info = demisto.getFilePath("825@9da8d636-cf30-42c2-8263-d09f5268be8a")
+        files = {'file': (file_info['name'], open(file_info['path'], 'rb'))}
+        access_token = get_access_token()
+        api_call_headers = {'Authorization': 'Bearer ' + access_token}
+        requests.post(API_URL + '/attachments', headers=api_call_headers, verify=False, files=files)
+    except Exception:
+        return_error("Entry does not contain a file.")
+
+    # return_error(res.status_code)
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: remove this comment

File: Integrations/ThreatQ/ThreatQ.py [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+        "type_id": "2",
+        "malware_locked": "0",
+        "tlp": {
+            "name": "GREEN"
+        }
+    }
+
+    res = tq_post_request('/attachments', params)
+    if res.status_code >= 400:
+        return_error('Could not upload file.')'''
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: remove lines 762-774

File: Integrations/ThreatQ/ThreatQ.py [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+    validate_ioc(email, 'email')
+    generic = {"Address": email}
+    raw, dbot = get_ioc_reputation(email, 'email')
+    entry_context = set_ioc_entry_context('email', raw, dbot, generic)
+    readable = build_readable("Search results for email address {0}".format(email), "indicator", raw)
+    return_outputs(readable, entry_context, raw)
+
+
+def get_domain_reputation(domain):
+    # fmt = validate_ioc(domain, 'domain')  # todo: add validation
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: remove this line

File: Integrations/ThreatQ/ThreatQ.py [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+''' EXECUTION CODE '''
+handle_proxy()
+command = demisto.command()
+LOG('command is {0}'.format(demisto.command()))
+try:
+    args = demisto.args()
+    if command == 'test-module':
+        test_module()
+    elif command == 'tq-search-by-name':
+        search_by_name_command(**args)
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: don't send **args, send nothing and handle getting the argument values in the command functions.
This is the Demisto convention.
This also applies to all the other functions in this if-else.

File: Integrations/ThreatQ/ThreatQ.py [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+
+''' EXECUTION CODE '''
+handle_proxy()
+command = demisto.command()
+LOG('command is {0}'.format(demisto.command()))
+try:
+    args = demisto.args()
+    if command == 'test-module':
+        test_module()
+    elif command == 'tq-search-by-name':
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: the commands should start with "threatq".
This also applies to all the other commands

File: Integrations/ThreatQ/ThreatQ.py [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+    elif command == "email":
+        get_email_reputation(**args)
+    elif command == "url":
+        get_url_reputation(**args)
+    elif command == "file":
+        get_file_reputation(**args)
+
+except Exception as ex:
+    raise
+    return_error(ex)
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: Cast to str, we can't be sure it's an str

File: Integrations/ThreatQ/ThreatQ.py [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+        get_domain_reputation(**args)
+    elif command == "email":
+        get_email_reputation(**args)
+    elif command == "url":
+        get_url_reputation(**args)
+    elif command == "file":
+        get_file_reputation(**args)
+
+except Exception as ex:
+    raise
```
Comment (guykeller) [inline Integrations/ThreatQ/ThreatQ.py]: no need to raise anything here

**Comment by guykeller** [comment]
General comments (fix across all the integration):
1. Every error message should start with “Error: ” then the error
2. Usually, it’s better if the construction of the context with the DT appears directly in the return_outputs call. See SignalSciences for reference
3. Calls to “return_outputs” should be with named params
4. I think wherever you write “readable” (make_readable or variable names) you should, whenever appropriate, change naming to “human_readable”. would be clearer
5. Whenever you specify Args in a docstring, every sentence should end with a dot
6. Docstrings should only be added to functions that need further explanation because they are more complicated.
We should try to use it when there is no other way (clear and descriptive naming, dividing to several smaller functions etc), to help the developers who will work on the code in the future understand it.
7. Commands don’t have return values, functions that use “return_outputs” don’t have return values in general.
8. All URL suffixes should be declared as globals in the beginning of the code (at the “Globals” section). See SignalSciences for reference
9. Each command should be split into two parts:
A. X_command - does all necessary modifications on user input, sends them to the X function, than uses what returned to return the proper output
B. X - handles the request itself and it’s logic
For example:
unlink_objects_command
unlink_objects

**Comment by guykeller** [comment]
From line 433 and above on please apply the following general changes:
1. Naming revision and improvements
2. Splitting actions done on one condensed line to several lines, for readability purposes

on lines 1-432, I commented on each case where I thought such a change was relevant.

Moreover, please pay attention - some CR notes do not refer only to that specific line, but to similar lines and fixes late

**Comment by guykeller** [comment]
And one quick tip - whenever you change naming, change it everywhere that function/variable is used.
This is a common pitfall of CR fixes

**Comment by ronykoz** [comment]
@dantavori what's the status here?

File: Integrations/ThreatQ/ThreatQ.py [hunk 8]
```diff
@@ -0,0 +1,899 @@
...
+API_URL = SERVER_URL + "/api"
+API_TOKEN_URL = API_URL + "/token"
+CLIENT_ID = demisto.params()['client_id']
+EMAIL = demisto.getParam('credentials').get('identifier')
+PASSWORD = demisto.getParam('credentials').get('password')
+USE_SSL = not demisto.params().get('insecure', False)
+THRESHOLD = int(demisto.params().get('threshold'))
+
+REGEX_MAP = {
+    'ipv4': re.compile(ipv4Regex, regexFlags),
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Integrations/ThreatQ/ThreatQ.py]: in CommonServerPython.py, this is the name of the regex string for IPv4. Is there something else to use?

File: Integrations/ThreatQ/ThreatQ.py [hunk 9]
```diff
@@ -0,0 +1,899 @@
...
+    'ipv6': re.compile(r"", regexFlags),
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Integrations/ThreatQ/ThreatQ.py]: IPv6 has no regex in CommonServerPython... So this one still doesn't have a validation, will be fixed of course

File: Integrations/ThreatQ/ThreatQ.py [hunk 10]
```diff
@@ -0,0 +1,899 @@
...
+    7: "Command and Control",
+    8: "Anonymization",
+    9: "Exfiltration",
+    10: "Host Characteristics",
+    11: "Compromised PKI Certificate",
+    12: "Login Compromise",
+    13: "Incident"
+}
+
+HEADERS = {
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Integrations/ThreatQ/ThreatQ.py]: The function that uses HEADERS dictionary (build_readable) is generic for all types of objects, using 'obj_type' variable. Separating the dict will triple the code length in this function. We could discuss more about it if you'd like.

File: Integrations/ThreatQ/ThreatQ.py [hunk 11]
```diff
@@ -0,0 +1,899 @@
...
+
+ERRORS_MAP = {
+    SEARCH_OBJECT: "Object was not found.",
+    SEARCH_TWO_OBJS: "One of the objects was not found.",
+    SEARCH_PIVOT: "Objects are not linked.",
+    SEARCH_ATTRIBUTE: "Attribute was not found.",
+    UNLINK_OBJS: "Could not unlink the objects."
+}
+
+DIRECTORY_MAP = {
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Integrations/ThreatQ/ThreatQ.py]: Yes

**Comment by bakatzir** [comment]
@ronykoz @dantavori  guys please also make sure you're following the 5.0 context standart: #17949
https://github.com/demisto/content/tree/master/docs/context_standards
Dan I mentioned you to this integration in the issue, please fill the checkbox when you're done.

**Comment by ronykoz** [comment]
@dantavori is this relevant? if not please close this PR as i see you have also https://github.com/demisto/content/pull/4276

**Comment by yardensade** [comment]
Can we delete this branch?

**Pull request closed without merging**